common/metrics/src/tracing.rs

```diff
@@ -0,0 +1,215 @@
+
+    pub struct FindU64<'a> {
+        pub name: &'a str,
+        pub value: u64,
+    }
+
+    impl<'a> FindU64<'a> {
+        pub fn new(name: &'a str) -> Self {
+            Self { name, value: 0 }
+        }
+    }
+
+    // Looking for multiple values because the callsite might be passed as a literal which turns into an i64 for example.
+    impl<'a> Visit for FindU64<'a> {
+        fn record_u64(&mut self, field: &Field, value: u64) {
+            if field.name() == self.name {
+                self.value = value;
+            }
+        }
+
+        fn record_i64(&mut self, field: &Field, value: i64) {
+            if field.name() == self.name {
+                self.value = value as u64;
+            }
+        }
+
+        fn record_i128(&mut self, field: &Field, value: i128) {
+            if field.name() == self.name {
+                self.value = value as u64;
+            }
+        }
+
+        fn record_u128(&mut self, field: &Field, value: u128) {
+            if field.name() == self.name {
+                self.value = value as u64;
+            }
+        }
+
+        fn record_debug(&mut self, _field: &Field, _value: &dyn std::fmt::Debug) {}
+    }
+}
+
+/// Check that the field exist on a type; if it doesn't this won't compile.
+/// This ensures that we're mapping fields with the correct name.
+#[macro_export]
+macro_rules! check_field {
+    ($event_ty:ident :: $field:ident) => {{
+        if false {
+            #[allow(clippy::needless_update)]
+            let _event = $event_ty {
+                $field: Default::default(),
+                ..Default::default()
+            };
+        }
+    }};
+}
+
+/// Set a gague to an absolute value based on a field in an event.
+#[macro_export]
+macro_rules! set_gauge {
+    ($event:ident, $event_ty:ident :: $field:ident, $gauge:expr) => {
+        check_field!($event_ty::$field);
+        let mut fld = visitors::FindU64::new(stringify!($field));
+        $event.record(&mut fld);
+        $gauge.set(fld.value as i64);
+    };
+}
+
+/// Increment a counter by the value of a field in the event.
+#[macro_export]
+macro_rules! inc_counter {
+    ($event:ident, $event_ty:ident :: $field:ident, $counter:expr) => {
+        check_field!($event_ty::$field);
+        let mut fld = visitors::FindU64::new(stringify!($field));
+        $event.record(&mut fld);
+        $counter.inc_by(fld.value);
+    };
+}
+
+/// Produce the prefixed event name from the type name.
+#[macro_export]
+macro_rules! event_name {
+    ($event_ty:ident) => {
+        concat!("event::", stringify!($event_ty))
+    };
+}
+
+/// Call one of the macros that set values on a metric.
+#[macro_export]
+macro_rules! event_mapping {
+    ($op:ident, $event:ident, $event_ty:ident :: $field:ident, $metric:expr) => {
+        $op!($event, $event_ty::$field, $metric);
+    };
+}
+
+/// Match the event name to event DTO types and within the map fields to metrics.
+#[macro_export]
+macro_rules! event_match {
```

Comment:
Out of the macros here at least this one needs to be supported by examples in its docstrings. It was highlighted as difficult to understand even when the the usage was right next to it.